### PR TITLE
[RNMobile] Refactor: Simplify media flow redux migration

### PR DESCRIPTION
**Pull request by fluiddot**
## Description

This PR refactors the [Simplify media flow redux store changes PR](https://github.com/WordPress/gutenberg/pull/30123) and changes the way the last block inserted is calculated.

The idea is that when a block is inserted via the inserter menu, we include a source param (`inserter_menu`) as part of the meta data. This way we can know where the last block was inserted from and in this case, we can display the media picker when it was inserted from the inserter menu.

## How has this been tested?

I tested the different blocks that use media (`image`, `gallery` and `video`).

1. Insert the block
2. Observe that the media picker **is displayed**
3. Insert another block
4. Observe that the media picker **is displayed**
5. Undo all the actions and redo them
6. Observe that the media picker **is not displayed**
7. Duplicate a block
8. Observe that the media picker **is not displayed**
9. Copy a block and insert it via the inserter menu
10. Observe that the media picker **is displayed**

## Screenshots 

<details><summary><strong>Gallery block</strong></summary>

https://user-images.githubusercontent.com/14905380/112467796-c6b79500-8d67-11eb-868e-3bcecf6747c7.mp4

</details>

<details><summary><strong>Image block</strong></summary>

https://user-images.githubusercontent.com/14905380/112467808-ca4b1c00-8d67-11eb-8016-a01e9805bbe3.mp4

</details>

<details><summary><strong>Video block</strong></summary>

https://user-images.githubusercontent.com/14905380/112467813-cae3b280-8d67-11eb-8458-2790e0630577.mp4

</details>

## Types of changes

N/A

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/store/actions.js
Comment (fluiddot) [inline packages/block-editor/src/store/actions.js:559]: Looks like the `meta` argument wasn't included in this action, however it's part of the arguments of `insertBlocks`.

File: packages/block-editor/src/store/reducer.js
Comment (fluiddot) [inline packages/block-editor/src/store/reducer.js:1749]: `RESET_BLOCKS` action is triggered in different situations like for example when executing undo/redo actions. Since it's resetting the blocks, the last block inserted should be cleared, this would be similar to the `CLEAR_LAST_BLOCK_INSERTED` action.

File: packages/block-library/src/gallery/edit.js
Comment (fluiddot) [inline packages/block-library/src/gallery/edit.js:114]: The dependencies array of `useSelect` is used to memoize the `mapSelect`, so the same `mapSelect` is invoked on every state change unless the dependencies change. For this case, this doesn't help us to improve the performance because this code is still being executed on every store change, so I had to move this logic inside a `useMemo` hook.

File: packages/block-editor/src/store/actions.js
Comment (jd-alexander) [reply packages/block-editor/src/store/actions.js:559]: Indeed! Thanks for updating `insertBlock` to match the parameter specifications.

File: packages/block-editor/src/store/reducer.js
Comment (jd-alexander) [reply packages/block-editor/src/store/reducer.js:1749]: Understood. This is a more optimal approach that I prefer since it relies on an existing action that will allow us to emulate the `CLEAR_LAST_BLOCK_INSERTED` behavior. The less code the better 😄

File: packages/block-library/src/gallery/edit.js
Comment (jd-alexander) [reply packages/block-library/src/gallery/edit.js:114]: Good eye 😉 I am seeing the correlation to the diff on the previous PR that utilized `useSelect` when computing the `resizedImages`.

File: packages/block-editor/src/store/reducer.js
Comment (jd-alexander) [inline packages/block-editor/src/store/reducer.js]: Quick question here, in my implementation I passed an empty object to clear `lastBlockInserted` and here you are passing null. Is there a standard for this?

Comment (fluiddot) [reply packages/block-editor/src/store/reducer.js]: Oh good catch! Here we should return an empty object because that's the default value for this state. I'm going to push the change asap.

Comment (fluiddot) [reply packages/block-editor/src/store/reducer.js]: I fixed this issue in this [commit](https://github.com/WordPress/gutenberg/pull/30238/commits/1164a3338fa9347a52bffcc34283d0986daf2e2c).

File: packages/block-library/src/video/edit.native.js
Comment (jd-alexander) [inline packages/block-library/src/video/edit.native.js]: @fluiddot do we have to duplicate 
```
lastBlockInserted.clientId === clientId &&
lastBlockInserted.source === 'inserter_menu';
```
across all three components? Could it be placed within the [selector](https://github.com/WordPress/gutenberg/blob/1164a3338fa9347a52bffcc34283d0986daf2e2c/packages/block-editor/src/store/selectors.js#L2114) where clientId and source get passed in as params ? Let me know.

Comment (fluiddot) [reply packages/block-library/src/video/edit.native.js]: Yeah, that's a good idea, we could use the selector you created [`wasBlockJustInserted`](https://github.com/WordPress/gutenberg/blob/1164a3338fa9347a52bffcc34283d0986daf2e2c/packages/block-editor/src/store/selectors.js#L2097-L2106).

Comment (fluiddot) [reply packages/block-library/src/video/edit.native.js]: I've just updated the `wasBlockJustInserted` selector in this [commit](https://github.com/WordPress/gutenberg/pull/30238/commits/f08846a4d511354051859d74a56fc0fac89d92a8).

File: packages/block-library/src/video/edit.native.js
Comment (jd-alexander) [inline packages/block-library/src/video/edit.native.js]: Not a blocker but I just noticed that the logic here could be converted to one line.`wasBlockJustInserted: select( blockEditorStore).wasBlockJustInserted( clientId, 'inserter_menu' )`

**Review by jd-alexander — APPROVED** [review]
Thanks for the improvements here @fluiddot I tested the new approach to signaling the source of block insertion and it works as expected in the image, video and gallery block. LGTM 🚢 

Once this is merged into the base branch, I will make the relevant updates such as 

1. Removing the `CLEAR_LAST_BLOCK_INSERTED` actions, reducers, selectors. 
2. Removing the `CLEAR_LAST_BLOCK_INSERTED` tests. 
3. Creating tests for the `INSERT_BLOCKS` and `RESET_BLOCKS` action changes in the reducer.
4. Updating the `lastBlockInserted` selector tests.

File: packages/block-library/src/video/edit.native.js
Comment (fluiddot) [reply packages/block-library/src/video/edit.native.js]: Yep, I saw both patterns in different components so I guess it depends if you want to take multiple selectors from the same store. In this case since it's only one, I think we could reduce it to one line 👍 .

Comment (fluiddot) [reply packages/block-library/src/video/edit.native.js]: I made this change in in this [commit](https://github.com/WordPress/gutenberg/pull/30238/commits/289e50616b1133db3d433cd2c6ea15b882e59cc0).

**Comment by fluiddot** [comment]
> Thanks for the improvements here @fluiddot I tested the new approach to signaling the source of block insertion and it works as expected in the image, video and gallery block. LGTM 🚢
> 

Awesome! I'll merge it once the PR checks pass.

> Once this is merged into the base branch, I will make the relevant updates such as
> 
> 1. Removing the `CLEAR_LAST_BLOCK_INSERTED` actions, reducers, selectors.
> 2. Removing the `CLEAR_LAST_BLOCK_INSERTED` tests.
> 3. Creating tests for the `INSERT_BLOCKS` and `RESET_BLOCKS` action changes in the reducer.
> 4. Updating the `lastBlockInserted` selector tests.

Yeah, we should cleanup the code and add some unit tests, thanks 🙇 !

**Pull request merged**
